sandbox/grist/imports/import_csv.py

```diff
@@ -91,107 +90,116 @@ def parse_file(file_path, parse_options=None):
   parse_options = parse_options or {}
 
   with open(file_path, "rb") as f:
+    sample = f.read(100000)
```

Comment:
I've found no way to prevent opening the file twice, given csv module wants a file opened as `str` afaik, but we need to open it as bytes to guess the encoding.

messytables was doing two things to work around that:
- opening the file as bytes and passing a custom iterator to `csv.reader` that would decode each line on the fly
- storing "sample" rows in memory and iterating first in those rows before continuing with the other rows

May I try to replicate this behaviour ?

Reply:
Only this line needs to be in the `with` block. The `sample` variable can still be used after.

You could also pass `sample.decode(encoding, errors='ignore')` to `_guess_dialect` instead of a file object to avoid reading and seeking.

Reading the file twice seems fine.

Reply:
> You could also pass sample.decode(encoding, errors='ignore') to _guess_dialect instead of a file object to avoid reading and seeking.

With the current implementation, the called of `_guess_dialect` does not have access to the sample. `_parse_open_file` needs an already opened file (as says the name).
One option may be to call `_guess`dialect` in `parse_file`, and then to pass the dialect properties as `kwargs`. Should I try that ? It may affect the way we write tests.

Reply:
No, I think it's better to forget what I said, I didn't have the full picture in mind.